### PR TITLE
feat: add etcd storer

### LINKED ISSUE #7: feature: etcd support
Since APISiX configuration information is stored in etcd, we also need to have the ability to interact with etcd. We need to get the service discovery-related configuration items, which determine what the program should do. Since APISIX can dynamically change its configuration at runtime, we need to be able to read the data in etcd and watch it for updates. Finally, we also need to be able to modify the configuration entries in etcd so that APISIX can obtain the results of service discovery.

We can refer to the relevant implementation in APISIX-Dashboard: [
apisix-dashboard/storage](https://github.com/apache/apisix-dashboard/tree/master/api/internal/core/storage)

### PR DESCRIPTION
**Pull request by gu18168**
resolve #7

### CODE REVIEW

File: internal/core/entity/entity.go [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+func (u *Upstream) SetNodes(nodes []*Node) {
+	(u.UpstreamDef).SetNodes(nodes)
+}
+
+// --- structures for upstream end ---
+
+type Route struct {
+	BaseInfo
+	URI             string                 `json:"uri,omitempty"`
```
Comment (spacewander) [inline internal/core/entity/entity.go]: Better only keep the required fields.

File: internal/core/entity/entity.go [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+	Port     int         `json:"port,omitempty"`
+	Weight   int         `json:"weight"`
+	Metadata interface{} `json:"metadata,omitempty"`
+}
+
+type NodesSetter interface {
+	SetNodes([]*Node)
+}
+
+type Healthy struct {
```
Comment (spacewander) [inline internal/core/entity/entity.go]: Better only keep in-used structs

File: internal/utils/utils.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	}
+	_sf = sonyflake.NewSonyflake(sonyflake.Settings{
+		MachineID: func() (u uint16, e error) {
+			return sumIPs(ips) + salt, nil
+		},
+	})
+	if _sf == nil {
+		panic("sonyflake init failed")
+	}
+}
```
Comment (spacewander) [inline internal/utils/utils.go]: Need newline between functions

File: internal/core/entity/entity.go [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+	ID         string `json:"id"`
+	CreateTime int64  `json:"create_time,omitempty"`
+	UpdateTime int64  `json:"update_time,omitempty"`
+}
+
+func (info *BaseInfo) GetBaseInfo() *BaseInfo {
+	return info
+}
+
+func (info *BaseInfo) Creating() {
```
Comment (spacewander) [inline internal/core/entity/entity.go]: As this project only rewrites the etcd data, there is no need to provide a Create method. The snowflake dependency can be removed too.

File: internal/core/entity/entity.go (hunk 1 above)
Comment (gu18168) [reply internal/core/entity/entity.go]: One question I have here is that if we remove the non-required fields, then when we update etcd, the serialization of the struct will also be without the filed value, so will it overwrite the existing value in etcd?

File: internal/core/entity/entity.go (hunk 4 above)
Comment (gu18168) [reply internal/core/entity/entity.go]: I will remove some unnecessary code.

File: internal/core/entity/entity.go (hunk 1 above)
Comment (spacewander) [reply internal/core/entity/entity.go]: Can we find a way to solve it? There will be new fields added across versions. We need to keep the fields that can't be decoded.

Comment (gu18168) [reply internal/core/entity/entity.go]: In the latest code, I tried a way to remove unnecessary fields from the entities.

File: internal/core/entity/entity_test.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+	embedElm(reflect.ValueOf(&route), route.All)
+	expectedAll := map[string]interface{}{
+		"uris": `[/*]`,
+		"upstream": map[string]interface{}{
+			"discovery_type": "nacos",
+			"service_name":   "TEST-SERVICE",
+			"nodes":          nodes,
+		},
+	}
+	assert.True(t, reflect.DeepEqual(route.All["upstream"], expectedAll["upstream"]), "Test latest attribute")
```
Comment (spacewander) [inline internal/core/entity/entity_test.go]: Should we compare route.All with expectedAll?

File: internal/core/entity/entity_test.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+	embedElm(reflect.ValueOf(&service), service.All)
+	expectedAll := map[string]interface{}{
+		"name": "TEST",
+		"upstream": map[string]interface{}{
+			"discovery_type": "nacos",
+			"service_name":   "TEST-SERVICE",
+			"nodes":          nodes,
+		},
+	}
+	assert.True(t, reflect.DeepEqual(service.All["upstream"], expectedAll["upstream"]), "Test latest attribute")
```
Comment (spacewander) [inline internal/core/entity/entity_test.go]: ditto

File: internal/core/entity/entity_test.go (hunk 5 above)
Comment (gu18168) [reply internal/core/entity/entity_test.go]: I have made the changes.

**Review by spacewander — APPROVED** [review]
@gu18168 
Nice job!

**Pull request merged**